### ISSUE TITLE
How to add user binaries to rancheros

### ISSUE THREAD

**Issue by schmurfy** [rancher/os#2322]
**RancherOS Version: (ros os version)**
v1.3.0

**Where are you running RancherOS? (docker-machine, AWS, GCE, baremetal, etc.)**
baremetal booted using PXE

I have toying around with the build process and I managed to have a kernel with the modules I need running but what I am missing is how to add files to the base system.

I added a hook script in os-kernel to build the zfs module and at the same time the zfs binaries, I copied them in the dist folder to make them available after the build is complete but now with the `os` project I am a bit lost as to where I could add my binaries, there is no hook I can find and I would rather not rebuild `os-base ` from scratch if I cna avoid it.

I am not clear where the base system is even stored for pxe between vmlinuz and initrd but it does not really matter, what I was trying to find is where I could add my files after os-base was added into the docker but Ifailed so far.

**Comment by niusmallnan** [comment]
We use the `buildroot` to build our `rootfs`. `https://github.com/rancher/buildroot`
We keep the build config in `os-base`. `https://github.com/rancher/os-base`
If users build their own `os-base`, they can update `OS_BASE_URL` in the dapper file of `rancher/os`.


In addition, if `os-base` cannot meet special requirements, they can change the default console.
`https://github.com/rancher/os/tree/master/images/02-console`
If they want to add files in other consoles too, they need to focus on here.
`https://github.com/rancher/os-services/tree/master/images`


If these files need to be used during the bootloader process, they should focus on initrd.
`https://github.com/rancher/os/blob/master/scripts/layout-initrd`

**Comment by schmurfy** [comment]
Thanks for the pointers, I think I got it but there is still one thing I am not clear about: what is a "console" for rancheros ?
It looks like this is where I end up when connecting using ssh, does that mean the ssh session is running inside a docker container ?

**Comment by niusmallnan** [comment]
Yes, RancherOS can switch consoles,  like default(busybox), ubuntu, centos, ...

The sshd daemon runs in these consoles, each console is a container.

**Comment by schmurfy** [comment]
nevermind I found what I was looking for, I added the binaries to the bootstrap container and changed the bootstrap go code to run `zfs import -a` & `zfs mount -a`, for some reasons I could not make it work using bootcmd kernel parameters.
I use kernel parameters to set the storage_driver to zfs and the graph path to the correct mount path.

I have one issue left which is weird, `zfs mount -a` fails because files already exist in the mount point, it is as if the `os-docker` system-docker was started before the `os-bootstrap` but judging by what is shown in the boot console this isnot the case and I don't understand why docker would complain that the filesystem is not correct (since it is not zfs) but still create some empty folders there.
